Use multipart uploads for large writes in PrestoS3FileSystem

ZacBlanco opened the pull request:
CTAS without partitioning will, for each node, basically write one huge file to the local disk before uploading it when the write it complete. This change attempts to upload as the write it occurring to prevent excessive disk usage.

## Description
<!---Describe your changes in detail-->

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Connector Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

ClarenceThreepwood commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -793,6 +797,44 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
+            assertEquals(s3.initiateMultipartRequests, 1);
+            assertEquals(s3.uploadPartRequests, 2);
+            assertEquals(s3.completePartUploadRequests, 1);
+        }
```
Maybe add a test that simulates error and subsequent cleanup?

ZacBlanco replied:
Added

ZacBlanco commented on presto-hive-common/src/main/java/com/facebook/presto/hive/HiveCommonSessionProperties.java:
```diff
@@ -176,6 +179,11 @@ public HiveCommonSessionProperties(HiveCommonClientConfig hiveCommonClientConfig
+                dataSizeProperty(
+                        S3_MIN_MULTIPART_PART_SIZE,
+                        "minimum upload size for multipart uploads to s3",
+                        io.airlift.units.DataSize.succinctDataSize(100, DataSize.Unit.MEGABYTE),
```
It's already configurable within the hive connector. I just exposed it as a session parameter for benchmarking purposes. I am fine removing it if you don't think it should be a session parameter

ClarenceThreepwood replied:
NM - I was looking in the wrong place. This is in HiveS3Config

However the default there appears to be 5MB. Should we make these consistent?

ZacBlanco replied:
I agree on making them consistent. However, S3 also enforces a 5MB minimum part limit with a maximum of 10k parts. This gives a default maximum file size of ~50G. Given that with this new implementation we don't know the total file size ahead of time, 50G seems like a relatively low limit, at least on the scale of Presto :)

I did intentionally set this to a larger value to allow for larger file sizes - 100MiB gives a 20x increase to the 50G limit for a max upload of ~1T, with little downsides IMO. Larger part sizes should generally decrease overhead due to having to make lest S3 requests.

I think it is better to make the default minimum part size larger than the current 5MiB. 100MiB was mostly an arbitrary pick to lift the file limit, but I am open to other suggestions.

majetideepak replied:
For reference, we use 10MiB in Velox. That would give a 100GiB file size.
Velox also did not make the part size configurable. It helps to minimize the configs. If required, 
we can make the part size adaptive as we upload parts. Arrow does this.
Each presto writer will consume 100MiB. For CTAS queries with many writers, will there be a higher change of OOM?

ZacBlanco replied:
The 10MiB reference from Velox is useful. I am okay aligning on that value for now just for consistency's sake. I also like the idea of using an adaptive approach. There are probably a few ways to do it - I think that's better left for a follow-up PR though.

As for the memory portion. 100MiB isn't necessarily used by each writer. When writing, the data first goes to a temp file on disk through a `BufferedInputStream` which I believe uses an 8K buffer by default. Then, when uploading through the S3 client, you specify the temp file on disk. The S3 client takes care of streaming it through it's own input stream. I dug through the S3 client code and didn't see anywhere that loads the part file all at once to be uploaded. It seems to be entirely streaming, so I think that memory usage/OOM is a concern regardless of part size.

majetideepak replied:
If the upload is streaming from a file, then the memory usage should be low. The overhead should be similar to the TransferManager path?

ZacBlanco replied:
Sorry - typo in my original comment. I meant to say it _is not_ a concern regardless of part size.

ZacBlanco replied:
I've updated the configs in both places to default to 10MiB.

ZacBlanco commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java:
```diff
@@ -1288,25 +1350,97 @@ public void close()
+                    newTempFileBuffer();
+                }
+            }
+            catch (SdkClientException | IOException e) {
```
Good point. Originally, I was trying to mimic the error handling from the previous code, however I have given it more thought I think it does make sense to have a catch-all block here and abort regardless of the exception. This has been fixed

ClarenceThreepwood reviewed (dismissed):
lgtm

aaneja commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java:
```diff
@@ -1288,25 +1349,97 @@ public void close()
 
-        private void uploadObject()
-                throws IOException
+        private void flushPartUpload(boolean last)
```
With this implementation, we are not utilizing the disk as a buffer adequately IMO. The uploads follow a create-part->upload-part-> delete-part flow, which means that our overall write speed is bounded by the network speed/bw (or the disk speed).
We could address this by creating a pipelined writer where the upload and delete of a part happens parallelly in the background and while the part-writer produces parts in a different thread. We can cap the total number of on-disk parts so that we don't fill up the disk.

ZacBlanco replied:
I had considered something like this as an option. However, I felt the pipelined version was too complex to pursue. The main goal for this PR was just to lift the disk-size limitation. Additional performance gain was a non-goal, but we happened to get some by properly overriding the `FilterOutputStream`

ZacBlanco replied:
With the results [here](https://github.com/prestodb/presto/pull/22424#issuecomment-2052150813), I am of the opinion that we should let this PR merge since the performance is always better than master in the benchmark configurations. If we want to improve perf further, we can iterate in another PR on a proper streaming implementation. Trino also has a proper streaming implementation.

When implementing the pipelining I think there are more design decisions that we may wish to make (or further configuration) which is not really in scope for the goal of the PR, which was lifting the disk space limitation.

aaneja replied:
Sounds good. Can you make an issue and link as a TODO for the streaming implementation ?

ZacBlanco replied:
Created https://github.com/prestodb/presto/issues/22536

aaneja commented:
>The additional performance gain is due to the fact that the original implementation relied on extending FilterOutputStream. In the default implementation of the FilterOutputStream, all writes end up occurring through the write(byte b) interface

Could you measure what the impact would be be if we fixed this problem only without replacing the TransferManager ? I'm also curious if you played around with [`task.writer-count`](https://github.com/prestodb/presto/blob/641c2693879270996ce463368b0d6afe0b636338/presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java#L437-L458) (and `task.partitioned-writer-count`) to see if we can utilize the disk/network better. I found this [old thread](https://groups.google.com/g/presto-users/c/SooEm5B203c) which discusses the impact of these configs

aaneja commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -793,6 +800,109 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
     }
 
+    @Test
+    public void testMultipart()
```
Can you also add a test that writes a complete pre-canned file using the PrestoS3FileSystem and reads it back and asserts on its contents ? Testing around boundary cases (under min part size, just above min part size, large file size) will prove robustness

ZacBlanco replied:
Added here: https://github.com/prestodb/presto/blob/ec82b5ea7861e87960288fae99956e3ffde98787/presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java#L910

ZacBlanco commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java:
```diff
@@ -1288,25 +1349,97 @@ public void close()
+
+        private String uploadSingleFile()
+        {
+            TransferManager transfer = TransferManagerBuilder.standard()
```
It could also be that the min part size was configured to a very large size. I didn't dig into what optimizations the transfer manager might make for larger-sized files. I just deferred to that implementation since it seems robust enough to handle large file uploads

ZacBlanco replied:
With the results [here](https://github.com/prestodb/presto/pull/22424#issuecomment-2052150813), I am of the opinion that we should let this PR merge since the performance is always better than master in the benchmark configurations. If we want to improve perf further, we can iterate in another PR on a proper streaming implementation. Trino also has a proper streaming implementation.

When implementing the pipelining I think there are more design decisions that we may wish to make (or further configuration) which is not really in scope for the goal of the PR, which was lifting the disk space limitation.

ZacBlanco commented:
@aaneja I re-ran the benchmarks for creating the `tpch.orders` orders table and built a branch which only overrides the outputstream methods. Additionally, I tested with `task.writer-count` set to the default (1) and 8. I also tested with a larger min part size for one of the runs on my PR. Here are. the results:

![image](https://github.com/prestodb/presto/assets/3464508/a67d6728-5707-4b8a-a390-84a98a3ba777)

aaneja commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -793,6 +803,179 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
+        try (FSDataInputStream in = fs.open(writtenFile)) {
+            assertEquals(actual.length, in.read(input));
+        }
+        assertEquals(actual, input, "failed on length " + actual.length);
```
The 'actual' here is the expected data

aaneja commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -793,6 +803,179 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
+        Path writtenFile = writerExpected.get();
+        byte[] input = new byte[actual.length];
+        try (FSDataInputStream in = fs.open(writtenFile)) {
+            assertEquals(actual.length, in.read(input));
```
We're missing the case where the actually written data is larger than expected. How about we use this for the method -
```
private static void assertRoundTripWrite(IOSupplier<Path> writerExpected, FileSystem fs, byte[] expected)
            throws IOException
    {
        Path writtenFile = writerExpected.get();
        byte[] actual;
        try (FSDataInputStream in = fs.open(writtenFile)) {
            actual = ByteStreams.toByteArray(in);
        }
        assertEquals(actual, expected);
    }
```

ZacBlanco replied:
Thanks - good catch. I've updated the method

aaneja commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/MockAmazonS3.java:
```diff
@@ -150,4 +216,81 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
     {
     }
+
+    @Override
```
Nice work with getting these added to the mock !

aaneja commented:
Aside : Your [experiment](https://github.com/prestodb/presto/pull/22424#issuecomment-2052150813) showed that the default `task.writer-count` may not be appropriate. Do you think we should set this higher ? Any suggestions on how we could set this dynamically for a better OOTB experience ?

ZacBlanco commented:
I did some digging into `task.writer-count`. I think the ideal value will depend on a number of hardware factors: the writer's core count, disk metrics (capacity, bandwidth, media type), network bandwidth. Additionally, the rate at which data can be fed into a `TableWriter` operator will determine the write speed. That rate could depend on a variety of non-hardware factors too such as the file format and compression algorithm. There could be more. Those are just the factors I could think of off the top of my head.

In these few experiments I ran without compression and with NVMe disks to try to achieve the highest data rate possible to stress the new implementation. We found that the performance increased, but for other workloads without compression I can't say that we would see gains as large. I could also see a situation which, if deployed using traditional spinning platter disks, concurrent writes would lead to abysmal performance.

I don't think there's a one-size fits all where we could guarantee better performance, but I think most "standard" use cases ("standard" referring to our standard cluster types used for benchmarking) would probably benefit from a writer count at least equal to the number of threads on the machine. Most deployments nowadays are probably using SSD and have multiple cores, so I think it's probably safe to increase the parallelism by a bit since it seems to give a measurable benefit. This does affect table layouts though (more files), so that's another thing to consider.

aaneja commented:
I think we need the task writer to be adaptive to scale out/in based on utilization levels. I'm not sure if Prestissimo does this either, I will check. We can suggest this improvement if it does not

ZacBlanco commented:
@tdcmeehan do you want to review this? Or know anyone else who might want to?

jaystarshot commented:
Nice! But can this increase aws operation cost if the buffer size is too small?

ZacBlanco commented:
@jaystarshot You're correct that it can increase the costs by a fair margin. The minimum part size for S3 is 5MB. I put together a spreadsheet with the cost calculations for uploading 1PB to S3 with a fixed file size of 256MB and varying min part sizes:

<img width="785" alt="image" src="https://github.com/prestodb/presto/assets/3464508/e9ce4dc4-ed6d-4ed0-9360-ed7cc13279ba">

When using a small part size, the cost for uploading to S3 can be significantly higher than standard PUT requests. However, you also need to consider that previously, we deferred to the TransferManager class, which makes the decision on whether or not an upload should be a PutObject or Multipart, so I'm not 100% sure how often we already do multipart uploads with the existing implementation.

Additionally, if you're uploading data at this scale, the monthly storage cost is going to far outstrip the cost of uploading the data in the first place. If you were to store 1PB in S3 and use the minimum part size of 5MB, the one-time upload cost is just 5.13% of the monthly storage cost, linearly decreasing with increases in part size.

The data from the chart is in the following spreadsheet: https://docs.google.com/spreadsheets/d/1DNNpfbahHAiwcrBnhxaGlgexAhraVV9_yu9VUpxez3E/edit?usp=sharing

jaystarshot commented:
Sounds good, but for temporary tables it might make more sense to have a bigger buffer size.

tdcmeehan commented:
I think this warrants a release note, since there are consequences to one's S3 costs.  We haven't included a flag, so there's no way to back out of this change--is that OK?

ZacBlanco commented:
Users could "back out" of the increased costs by increasing the min-multipart-upload size configuration. If you'd prefer a boolean flag to flip to the old behavior I can do so.

I will add in a release note for it

tdcmeehan commented:
I think it's safer to introduce a schedule, where we introduce the change with a release note, then follow up with a change in flag default value, then deprecate the old functionality.  I'm just not confident that this won't have outsized impact on certain environments, and better visibility can help to mitigate that.

ZacBlanco commented:
I've updated the PR to add a config `presto.s3.use-streaming-uploads` which uses the new implementation. I split out the Input/OutputStream classes from the `PrestoS3FileSystem` file to make the class smaller and easier to read. The new implementation lives in `PrestoS3StreamingOutputStream`. Also, the "old" PrestoS3OutputStream implementation now overrides the `write` methods properly without changing any other behavior to get some performance wins by default.

steveburnett commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -411,7 +411,13 @@ Property Name                         Description
 ``hive.s3.multipart.min-file-size``   Minimum file size before multi-part upload to S3 is used.   ``16 MB``
 
-``hive.s3.multipart.min-part-size``   Minimum multi-part upload part size.                        ``5 MB``
+``hive.s3.multipart.min-part-size``   Minimum multi-part upload part size. Can also be set via    ``10MB``
```
```suggestion
``hive.s3.multipart.min-part-size``   Minimum multi-part upload part size. Can also be set using  ``10MB``
```
Do not use Latin abbreviations. For context and suggestions, see the GitLab [recommended word list entry for via](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#via). 

(I tested this, and discovered I needed to delete two blank spaces between "using" and "``10MB``" or the extra characters in "using" threw off the format for the Default value column.)

ZacBlanco replied:
fixed, thanks!

steveburnett commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -411,7 +411,13 @@ Property Name                         Description
 
-``hive.s3.multipart.min-part-size``   Minimum multi-part upload part size.                        ``5 MB``
+``hive.s3.multipart.min-part-size``   Minimum multi-part upload part size. Can also be set via    ``10MB``
+                                      the session property ``s3_multipart_min_part_size``
```
```suggestion
                                      the session property ``s3_multipart_min_part_size``.
```
Nit, punctuation.

ZacBlanco replied:
fixed

steveburnett reviewed (changes requested):
Looks good! Two nits, nothing major.

steveburnett reviewed (approved):
LGTM! (docs)

Pull updated branch, new docs build, docs look good. Thanks!

tdcmeehan commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3InputStream.java:
```diff
@@ -0,0 +1,349 @@
+                        }
+                        catch (RuntimeException e) {
+                            stats.newGetObjectError();
+                            if (e instanceof AmazonS3Exception) {
```
Please move this into a dedicated catch clause?

ZacBlanco replied:
I think it was originally written this way to only have to call `stats.newGetObjectError()` in only one catch branch to prevent mis-collecting stats. But I can put it in both catch clauses

ZacBlanco replied:
updated

tdcmeehan commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3InputStream.java:
```diff
@@ -0,0 +1,349 @@
+                    .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                    .stopOn(InterruptedException.class, PrestoS3FileSystem.UnrecoverableS3OperationException.class, EOFException.class, FileNotFoundException.class, AbortedException.class)
+                    .onRetry(stats::newGetObjectRetry)
+                    .run("getS3Object", () -> {
```
I would move this lambda into a new method and use a method reference.

ZacBlanco replied:
updated

tdcmeehan reviewed (commented):
LGTM just a couple nits

elharo commented on presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3InputStream.java:
```diff
@@ -0,0 +1,363 @@
+    @Override
+    public int read()
+    {
+        // This stream is wrapped with BufferedInputStream, so this method should never be called
```
This is a public class so this is not guaranteed.

ZacBlanco replied:
This a direct refactor from the original class which was private. I've removed the comment

elharo commented on presto-hive/src/test/java/com/facebook/presto/hive/s3/TestPrestoS3FileSystem.java:
```diff
@@ -794,6 +806,191 @@ public ObjectListing listObjects(ListObjectsRequest listObjectsRequest)
+    private void assertDirEmpty(java.nio.file.Path path)
+            throws IOException
+    {
+        try (Stream<java.nio.file.Path> dirIter = Files.walk(path)) {
```
This is more complex than it has to be. You can just do path.toFile().list()

ZacBlanco replied:
Thanks for the pointer, fixed
